src/groovy/heim/session/SessionCollector.groovy

```diff
@@ -0,0 +1,94 @@
+package heim.session
+
+import groovy.util.logging.Log4j
+import heim.tasks.JobTasksService
+import heim.tasks.TaskState
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+/**
+ * Created by piotrzakrzewski on 22/10/15.
+ */
+@Log4j
+@Component
+class SessionCollector {
+
+    def SessionCollector(){
+        new Timer().schedule({
+            garbageCollection()
+        } as TimerTask, collectionInterval, collectionInterval)
+    }
+
+    @Autowired
+    private SessionService sessionService
+
+    @Autowired
+    private JobTasksService jobTasksService
+
+
+    private final Integer collectionInterval = 3 * 60 * 1000 // 3 min
+    private final Integer sessionLifeSpan = 10 * 60 * 1000 // 10 min
+
+    private HashMap<UUID,Date> sessionLastActivity = new HashMap<UUID,Date>()
+    private HashMap<UUID,List> sessionTasks = new HashMap<UUID,List>()
+
+    void touchSession(UUID sessionId, UUID taskId){
+        registerTask(sessionId, taskId)
+        Date now = new Date()
+        sessionLastActivity.put(sessionId, now)
+        log.debug("Session: ${sessionId} lastActivity updated.")
+    }
+
+    void touchSession(UUID sessionId){
+        sessionLastActivity.put(sessionId, now)
+        log.debug("Session: ${sessionId} lastActivity updated.")
+    }
+
+    void registerTask(UUID sessionId, UUID taskId){
+        if(!sessionTasks.containsKey(sessionId)){
+            sessionTasks.put(sessionId,[])
+        }
+        sessionTasks.get(sessionId).add(taskId)
+    }
+
+    void garbageCollection(){
+        def terminatedSessions = []
+        sessionLastActivity.each {
+            sessionId, startTime ->
+                if(isStale(sessionId,startTime)){
+                    sessionService.destroySession(sessionId)
+                    terminatedSessions.add(sessionId)
+                    log.debug("Terminated session: ${sessionId} due to inactivity.")
+                }
+        }
+        terminatedSessions.each {
+            sessionId ->
+             sessionLastActivity.remove(sessionId)
+        }
+    }
+
+    boolean isStale(UUID sessionId, Date lastTouched){
+        Date now = new Date()
+        def delta = now.time() - lastTouched.time()
+        delta > sessionLifeSpan && noActiveTasks(sessionId)
+    }
+
+    boolean noActiveTasks(UUID sessionId){
+        def tasks = sessionTasks[sessionId]
+        boolean noActiveTasks = true
+        tasks.each {taskId ->
+            TaskState state = jobTasksService.getTaskState(taskId)
+            if(isActive(state)){
+                noActiveTasks = false
+            }
+            if(state.name() == TaskState.QUEUED){
+                touchSession(sessionId) // If session has queued jobs - refresh its lifespan. Let it actually do smt!
+            }
+        }
+        return noActiveTasks
+    }
+
+    boolean isActive(TaskState state){
+        !(state.name() == TaskState.FINISHED)
+    }
+}
```

Comment:
Mind the whitespace (note spaces after keywords such as `if`, `for` and after the function's arguments):

```
boolean isThat() {
    if (foobar) {
        doThat();
    }
}
```